crates/stateless/src/witness_db.rs

```diff
@@ -0,0 +1,145 @@
+    fn basic(&mut self, address: Address) -> Result<Option<AccountInfo>, Self::Error> {
+        let hashed_address = keccak256(address);
+        trace!(target: "reth-stateless::evm", %address, %hashed_address, "retrieving account");
+        let Some(bytes) = self.trie.get_account_value(&hashed_address) else {
+            trace!(target: "reth-stateless::evm", %address, %hashed_address, "no account found");
+            return Ok(None)
+        };
+        let account = TrieAccount::decode(&mut bytes.as_slice())?;
+        let account_info = AccountInfo {
+            balance: account.balance,
+            nonce: account.nonce,
+            code_hash: account.code_hash,
+            code: None,
+        };
+        trace!(target: "reth-stateless::evm", %address, %hashed_address, ?account_info, "account retrieved");
+        Ok(Some(account_info))
+    }
+
+    /// Get storage value of an account at a specific slot.
+    ///
+    ///  Returns `U256::ZERO` if the slot is not found in the trie.
+    fn storage(&mut self, address: Address, slot: U256) -> Result<U256, Self::Error> {
+        let slot = B256::from(slot);
+        let hashed_address = keccak256(address);
+        let hashed_slot = keccak256(slot);
+        trace!(target: "reth-stateless::evm", %address, %hashed_address, %slot, %hashed_slot, "retrieving storage slot");
+        let value = match self.trie.get_storage_slot_value(&hashed_address, &hashed_slot) {
+            Some(value) => U256::decode(&mut value.as_slice())?,
+            None => U256::ZERO,
+        };
+        trace!(target: "reth-stateless::evm", %address, %hashed_address, %slot, %hashed_slot, %value, "storage slot retrieved");
+        Ok(value)
+    }
+
+    /// Get account code by its hash from the provided bytecode map.
+    ///
+    /// Returns an error if the bytecode for the given hash is not found in the map.
+    #[cfg(not(feature = "ef-tests"))]
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        trace!(target: "reth-stateless::evm", %code_hash, "retrieving bytecode");
+        let bytecode = self.bytecode.get(&code_hash).ok_or_else(|| {
+            ProviderError::TrieWitnessError(format!("bytecode for {code_hash} not found"))
+        })?;
+
+        Ok(bytecode.clone())
+    }
+
+    #[cfg(feature = "ef-tests")]
+    fn code_by_hash(&mut self, code_hash: B256) -> Result<Bytecode, Self::Error> {
+        trace!(target: "reth-stateless::evm", %code_hash, "retrieving bytecode");
+        use alloy_eips::eip4788::BEACON_ROOTS_CODE;
+        let beacon_root_hash = keccak256(BEACON_ROOTS_CODE.clone());
+        if code_hash == beacon_root_hash {
+            return Ok(Bytecode::new_raw(BEACON_ROOTS_CODE.clone()));
+        }
+        let bytecode = self.bytecode.get(&code_hash).ok_or_else(|| {
+            ProviderError::TrieWitnessError(format!("bytecode for {code_hash} not found"))
+        })?;
+
+        Ok(bytecode.clone())
+    }
```

Comment:
This was added because the beacon root contract was not being included in the witness, when there were no transactions. 

I've put this here under the assumption that the "fix" would be to add it into the witness for post cancun blocks. We could also have the stateless code handle it, though it does special case the system contracts